### ISSUE TITLE
Apple MPS inference with detect.py

### ISSUE THREAD

**Issue by lightonthefloor** [ultralytics/yolov5#9596]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection

### Bug

When I run the command: python detect.py --source 0 --device mps, there is no detections. But when I run by using cpu, it has.

### Environment

- Yolo v5-6.2
- macOS 12.6
- conda 4.14
- python 3.10.4

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@lightonthefloor MPS inference is waiting on a few remaining operations to be integrated into torch nightly, so you should receive an error similar to this:

```python
python detect.py --source 0 --device mps

detect: weights=yolov5s.pt, source=0, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=mps, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v6.2-154-g5c34c4c Python-3.9.0 torch-1.12.1 MPS

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
1/1: 0...  Success (inf frames 1920x1080 at 15.00 FPS)

Traceback (most recent call last):
  File "/Users/glennjocher/PycharmProjects/yolov5/detect.py", line 255, in <module>
    main(opt)
  File "/Users/glennjocher/PycharmProjects/yolov5/detect.py", line 250, in main
    run(**vars(opt))
  File "/Users/glennjocher/PycharmProjects/yolov5/venv39/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/Users/glennjocher/PycharmProjects/yolov5/detect.py", line 126, in run
    pred = non_max_suppression(pred, conf_thres, iou_thres, classes, agnostic_nms, max_det=max_det)
  File "/Users/glennjocher/PycharmProjects/yolov5/utils/general.py", line 843, in non_max_suppression
    x = x[xc[xi]]  # confidence
NotImplementedError: The operator 'aten::index.Tensor' is not current implemented for the MPS device. If you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.

```

**Comment by lightonthefloor** [comment]
Nope. I have already set this environment variable.

**Comment by lightonthefloor** [comment]
`(yolo-v5_test) trisoil@Trisoils-MacBook-Pro yolov5-master % python detect.py --source 0 --device mps
detect: weights=yolov5s.pt, source=0, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=mps, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 🚀 2022-9-17 Python-3.10.4 torch-1.12.1 MPS

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
[ WARN:0@1.789] global /private/var/folders/nz/j6p8yfhx1mv_0grj5xl4650h0000gp/T/abs_c6qq9eqk9d/croots/recipe/opencv-suite_1663872527491/work/modules/videoio/src/cap_gstreamer.cpp (862) isPipelinePlaying OpenCV | GStreamer warning: GStreamer: pipeline have not been created
1/1: 0...  Success (inf frames 1920x1080 at 15.00 FPS)

/Users/trisoil/OneDrive - bupt.edu.cn/College Learning/Robocon/yolov5-master/utils/general.py:838: UserWarning: The operator 'aten::index.Tensor' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at  /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1659484612588/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  x = x[xc[xi]]  # confidence

(<unknown>:84503): GStreamer-CRITICAL **: 19:13:30.121: gst_element_make_from_uri: assertion 'gst_uri_is_valid (uri)' failed
[ WARN:0@3.850] global /private/var/folders/nz/j6p8yfhx1mv_0grj5xl4650h0000gp/T/abs_c6qq9eqk9d/croots/recipe/opencv-suite_1663872527491/work/modules/videoio/src/cap_gstreamer.cpp (2180) open OpenCV | GStreamer warning: cannot link elements
0: 384x640 (no detections), 167.8ms
0: 384x640 (no detections), 17.8ms
0: 384x640 (no detections), 14.6ms
0: 384x640 (no detections), 15.4ms
`

**Comment by glenn-jocher** [comment]
@lightonthefloor I'm getting a different error message if I use the fallback. Do you see the same thing?
```bash
PYTORCH_ENABLE_MPS_FALLBACK=1 python detect.py --device mps
```
<img width="1421" alt="Screenshot 2022-09-26 at 13 42 29" src="https://user-images.githubusercontent.com/26833433/192267903-c5a4b639-c846-4357-b0c8-d33554e80dca.png">

**Comment by lightonthefloor** [comment]
No. And I am not using the latest nightly. My torch version is 1.12.1

**Comment by glenn-jocher** [comment]
@lightonthefloor if you really want to use MPS today you can add this fix. This eliminates the need to pass the fallback flag, but it passes all predictions to CPU before NMS, which also takes time. My metrics on an M2 MacBook with this change are here. Note also that MPS is slow for the first image and then faster on subsequent images of the same size.

<img width="539" alt="Screenshot 2022-09-26 at 13 51 52" src="https://user-images.githubusercontent.com/26833433/192269688-92d995f8-102e-4c44-99fb-838348d313d4.png">

```bash
detect: weights=yolov5s.pt, source=data/images, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=mps, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v6.2-171-gbd9c0c4 Python-3.10.6 torch-1.13.0.dev20220926 MPS

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
image 1/3 /Users/glennjocher/PycharmProjects/yolov5/data/images/bus.jpg: 640x480 9 persons, 2 buss, 143.4ms
image 2/3 /Users/glennjocher/PycharmProjects/yolov5/data/images/zidane.jpg: 384x640 5 persons, 5 ties, 189.1ms
video 3/3 (1/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 184.8ms
video 3/3 (2/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 12.8ms
video 3/3 (3/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 11.2ms
video 3/3 (4/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 13.9ms
video 3/3 (5/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 16.7ms
video 3/3 (6/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 8.4ms
video 3/3 (7/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 8.5ms
video 3/3 (8/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 7.8ms
video 3/3 (9/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 7.6ms
video 3/3 (10/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 8.1ms
video 3/3 (11/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 7.6ms
video 3/3 (12/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 8.1ms
video 3/3 (13/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 3 persons, 7.7ms
video 3/3 (14/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 3 persons, 8.0ms
video 3/3 (15/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 3 persons, 11.3ms
video 3/3 (16/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 3 persons, 7.8ms
video 3/3 (17/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 7.5ms
video 3/3 (18/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 8.1ms
video 3/3 (19/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 2 trucks, 8.0ms
video 3/3 (20/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 11 persons, 8.0ms
video 3/3 (21/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 9 persons, 7.8ms
video 3/3 (22/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 9 persons, 7.9ms
video 3/3 (23/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 10 persons, 7.7ms
video 3/3 (24/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 7.9ms
video 3/3 (25/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 7.8ms
video 3/3 (26/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 3 persons, 8.0ms
Speed: 1.2ms pre-process, 26.5ms inference, 9.7ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs/detect/exp27
```

**Comment by glenn-jocher** [comment]
Equivalent CPU times looks like this. Faster NMS but much slower inference.

```bash
detect: weights=yolov5s.pt, source=data/images, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v6.2-171-gbd9c0c4 Python-3.10.6 torch-1.13.0.dev20220926 CPU

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
image 1/3 /Users/glennjocher/PycharmProjects/yolov5/data/images/bus.jpg: 640x480 4 persons, 1 bus, 161.2ms
image 2/3 /Users/glennjocher/PycharmProjects/yolov5/data/images/zidane.jpg: 384x640 2 persons, 2 ties, 130.0ms
video 3/3 (1/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 139.6ms
video 3/3 (2/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 151.8ms
video 3/3 (3/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 136.5ms
video 3/3 (4/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 134.6ms
video 3/3 (5/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 145.0ms
video 3/3 (6/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 131.3ms
video 3/3 (7/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 136.9ms
video 3/3 (8/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 166.2ms
video 3/3 (9/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 131.6ms
video 3/3 (10/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 135.5ms
video 3/3 (11/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 160.2ms
video 3/3 (12/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 145.7ms
video 3/3 (13/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 143.7ms
video 3/3 (14/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 143.1ms
video 3/3 (15/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 176.4ms
video 3/3 (16/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 179.0ms
video 3/3 (17/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 154.8ms
video 3/3 (18/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 176.2ms
video 3/3 (19/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 truck, 164.1ms
video 3/3 (20/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 5 persons, 152.0ms
video 3/3 (21/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 3 persons, 147.0ms
video 3/3 (22/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 3 persons, 209.0ms
video 3/3 (23/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 4 persons, 173.6ms
video 3/3 (24/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 109.8ms
video 3/3 (25/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 (no detections), 104.3ms
video 3/3 (26/26) /Users/glennjocher/PycharmProjects/yolov5/data/images/IMG_2590_rotate180.MOV: 640x384 1 person, 167.0ms
Speed: 0.6ms pre-process, 150.2ms inference, 0.8ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs/detect/exp40
```

**Comment by glenn-jocher** [comment]
@lightonthefloor good news 😃! Your original issue may now be fixed ✅ in PR #9600. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by DP1701** [comment]
@glenn-jocher 
Hi, have you tried the new nighlty build from pytorch for inference with mps?

Inference with torch 2.0.0, torchvision 0.15.1
<img width="1282" alt="2" src="https://user-images.githubusercontent.com/68444979/230744539-14465077-5ac7-4120-963a-b85610804196.png">
It just didn't work with a custom yolov5m detector

Inference with torch 2.1.0.dev20230408, torchvision 0.16.0.dev20230408 
<img width="1210" alt=" 1" src="https://user-images.githubusercontent.com/68444979/230744543-ea602fcd-f2fa-4526-ae85-27007e9644ab.png">
But now yes. At first glance, the detection fields seem to be all correct for me as well.
